core/src/main/java/com/graphhopper/routing/ch/CHPreparationGraph.java

```diff
@@ -0,0 +1,870 @@
+            return currEdge == null ? "not_started" : getBaseNode() + "-" + getAdjNode();
+        }
+
+        private boolean nodeAisBase() {
+            // in some cases we need to determine which direction of the (bidirectional) edge we want
+            return currEdge.getNodeA() == node;
+        }
+    }
+
+    interface PrepareEdge {
+        boolean isShortcut();
+
+        int getPrepareEdge();
+
+        int getNodeA();
+
+        int getNodeB();
+
+        double getWeightAB();
+
+        double getWeightBA();
+
+        int getOrigEdgeKeyFirstAB();
+
+        int getOrigEdgeKeyFirstBA();
+
+        int getOrigEdgeKeyLastAB();
+
+        int getOrigEdgeKeyLastBA();
+
+        int getSkipped1();
+
+        int getSkipped2();
+
+        int getOrigEdgeCount();
+
+        void setSkipped1(int skipped1);
+
+        void setSkipped2(int skipped2);
+
+        void setWeight(double weight);
+
+        void setOrigEdgeCount(int origEdgeCount);
+    }
+
+    public static class PrepareBaseEdge implements PrepareEdge {
+        private final int prepareEdge;
+        private final int nodeA;
+        private final int nodeB;
+        private final float weightAB;
+        private final float weightBA;
+
+        public PrepareBaseEdge(int prepareEdge, int nodeA, int nodeB, float weightAB, float weightBA) {
+            this.prepareEdge = prepareEdge;
+            this.nodeA = nodeA;
+            this.nodeB = nodeB;
+            this.weightAB = weightAB;
+            this.weightBA = weightBA;
+        }
+
+        @Override
+        public boolean isShortcut() {
+            return false;
+        }
+
+        @Override
+        public int getPrepareEdge() {
+            return prepareEdge;
+        }
+
+        @Override
+        public int getNodeA() {
+            return nodeA;
+        }
+
+        @Override
+        public int getNodeB() {
+            return nodeB;
+        }
+
+        @Override
+        public double getWeightAB() {
+            return weightAB;
+        }
+
+        @Override
+        public double getWeightBA() {
+            return weightBA;
+        }
+
+        @Override
+        public int getOrigEdgeKeyFirstAB() {
+            int keyFwd = prepareEdge << 1;
+            if (nodeA > nodeB)
+                keyFwd += 1;
+            return keyFwd;
+        }
+
+        @Override
+        public int getOrigEdgeKeyFirstBA() {
+            int keyBwd = prepareEdge << 1;
+            if (nodeB > nodeA)
+                keyBwd += 1;
+            return keyBwd;
+        }
+
+        @Override
+        public int getOrigEdgeKeyLastAB() {
+            int keyFwd = prepareEdge << 1;
+            if (nodeA > nodeB)
+                keyFwd += 1;
+            return keyFwd;
+        }
+
+        @Override
+        public int getOrigEdgeKeyLastBA() {
+            int keyBwd = prepareEdge << 1;
+            if (nodeB > nodeA)
+                keyBwd += 1;
+            return keyBwd;
+        }
```

Comment:
Why do we need those 4 methods? The methods getOrigEdgeKeyLastBA and getOrigEdgeKeyFirstBA
and then getOrigEdgeKeyFirstAB and getOrigEdgeKeyLastAB look identical and could we just omit them as this isn't used elsewhere? Also instead of
```
if (nodeA > nodeB) keyFwd += 1;
return keyFwd;
```

we could write:
```
return nodeA > nodeB? keyFwd + 1 : keyFwd;
```
but probably this is done from the JIT already

Reply:
Ah, sorry. Forget this comment I just found another usage (EdgeBasedPrepareShortcut)...

Reply:
Ok I forgot it already :) (but I made use of the ternary operator as its a bit more concise).